engine/client/secretprovider/vault.go

```diff
@@ -3,20 +3,44 @@ package secretprovider
 import (
 	"context"
 	"fmt"
+	"net/url"
 	"os"
 	"strings"
+	"time"
 
 	vault "github.com/hashicorp/vault/api"
 	auth "github.com/hashicorp/vault/api/auth/approle"
 )
 
+type dataWithTTL struct {
+	expiresAt time.Time
+	data      map[string]any
+}
+
 var (
 	vaultClient *vault.Client
-	vaultCache  = make(map[string]map[string]any)
+	vaultCache  = make(map[string]dataWithTTL)
 )
 
 // HashiCorp Vault provider for SecretProvider
-func vaultProvider(ctx context.Context, key string) ([]byte, error) {
+func vaultProvider(ctx context.Context, pathWithQuery string) ([]byte, error) {
+	parsed, err := url.Parse(pathWithQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// this is just path part without the query params such as ttl
+	key := parsed.Path
+
+	var ttl time.Duration
+	ttlStr := strings.TrimSpace(parsed.Query().Get("ttl"))
+	if ttlStr != "" {
+		ttl, err = time.ParseDuration(ttlStr)
+		if err != nil {
+			return nil, fmt.Errorf("invalid ttl %q provided for secret %q. err: %w", ttlStr, key, err)
```

Comment:
The standard way of doing this is:

```suggestion
			return nil, fmt.Errorf("invalid ttl %q provided for secret %q: %w", ttlStr, key, err)
```